File: automation-test-selenium/src/main/java/org/suntower/pages/components/BootstrapModalComponent.java
package org.suntower.pages.components;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.suntower.core.BasePage;

public class BootstrapModalComponent extends BasePage {
  private final By modal;

  public BootstrapModalComponent(WebDriver driver) {
    this(driver, ".modal.show");
  }

  public BootstrapModalComponent(WebDriver driver, String selector) {
    super(driver);
    this.modal = css(selector);
  }

  public By byId(String id) {
    return css("#" + id);
  }

  public By visible() {
    return modal;
  }

  public void waitForVisible() {
    waitForVisible(modal);
  }

  public void waitForVisible(By locator) {
    super.waitForVisible(locator);
  }

  public void waitForHidden() {
    waitForHidden(modal);
  }

  public void waitForHidden(By locator) {
    super.waitForHidden(locator);
  }

  public void close() {
    close(modal);
  }

  public void close(By locator) {
    By closeButton = css(selector(locator) + " .modal-header .btn-close, " + selector(locator) + " .modal-footer button, " + selector(locator) + " [data-bs-dismiss='modal']");
    click(closeButton);
    waitForHidden(locator);
  }

  public String text() {
    return text(modal).trim();
  }

  public String text(By locator) {
    return super.text(locator).trim();
  }

  private String selector(By locator) {
    String raw = locator.toString();
    if (!raw.startsWith("By.cssSelector: ")) {
      throw new IllegalArgumentException("BootstrapModalComponent requires CSS selectors.");
    }
    return raw.replace("By.cssSelector: ", "");
  }
}

File: automation-test-playwright/pages/components/BootstrapModalComponent.ts
import { type Locator, type Page } from "@playwright/test";
import { env } from "@helpers-runtime/env";

/**
 * Encapsulates interactions with Bootstrap modal dialogs.
 * Provides wait, close, and content extraction utilities for `.modal.show` elements.
 */
export class BootstrapModalComponent {
  readonly modal: Locator;

  /**
   * Creates a modal component. Defaults to the standard Bootstrap `.modal.show` selector.
   */
  constructor(
    private readonly page: Page,
    selector = ".modal.show"
  ) {
    this.modal = this.page.locator(selector);
  }

  byId(id: string): Locator {
    return this.page.locator(`#${id}`);
  }

  visible(): Locator {
    return this.modal;
  }

  async waitForVisible(locator: Locator = this.modal): Promise<void> {
    await locator.waitFor({ state: "visible", timeout: env.expectTimeout });
  }

  async waitForHidden(locator: Locator = this.modal): Promise<void> {
    await locator.waitFor({ state: "hidden", timeout: env.expectTimeout });
  }

  /**
   * Closes the page UI element.
   */
  async close(locator: Locator = this.modal): Promise<void> {
    await locator
      .locator(".modal-header .btn-close, .modal-footer button, [data-bs-dismiss='modal']")
      .filter({ visible: true })
      .first()
      .click();
    await this.waitForHidden(locator);
  }

  async text(locator: Locator = this.modal): Promise<string> {
    return ((await locator.innerText()) ?? "").trim();
  }
}
